### PR TITLE
Fix _analytical_jacobian to not require in-place grad accumulation

### PR DESCRIPTION
**Pull request by colesbury**
This doesn't fix any bugs, but makes the Jacobian tests robust against changes to `AccumulateGrad`. Previously, the Jacobian tests assumed that `AccumulateGrad` operates in-place, which is true, but an implementation detail.

### CODE REVIEW

File: test/common_nn.py [hunk 1]
```diff
@@ -564,21 +564,21 @@ def _zero_grad_input(self, input):
...
-            num_param = sum(p.numel() for p in param)
-            jacobian_param = torch.zeros(num_param, d_out.nelement())
+            num_param = sum(p.numel() for p in self._get_parameters(module)[0])
+            jacobian_param = torch.zeros(num_param, output_size)
 
-        for i in range(flat_d_out.nelement()):
-            d_out.zero_()
+        for i in range(output_size):
+            _, d_param = self._get_parameters(module)
+            d_out = torch.zeros_like(output_t)
```
Comment (apaszke) [inline test/common_nn.py]: I think you can still reuse `d_out`. It's not grad of any param; it's a tensor you created. The important part is reassigning `d_param` at every iteration

File: test/common_nn.py [hunk 2]
```diff
@@ -603,12 +603,6 @@ def _analytical_jacobian(self, module, input, jacobian_input=True, jacobian_para
         return res
 
     def _numerical_jacobian(self, module, input, jacobian_input=True, jacobian_parameters=True):
-        output = self._forward(module, input)
-        output_size = output.nelement()
-
-        if jacobian_parameters:
-            param, d_param = self._get_parameters(module)
```
Comment (apaszke) [inline test/common_nn.py]: This is just a code cleanup. No important fixes here, right?

**Review by apaszke — APPROVED** [review]
Looks good, but I think you could save on re-creating `d_out` every time.

File: test/common_nn.py (hunk 2 above)
Comment (colesbury) [reply test/common_nn.py]: Yes

File: test/common_nn.py (hunk 1 above)
Comment (colesbury) [reply test/common_nn.py]: Yes, this is stylistic: I think it's cleaner to re-create temporaries rather than calling `zero_()` every iteration.

**Pull request merged**